server/src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -224,31 +223,47 @@ private static <K, T, H> MapDiff<K, T, H> diffOrEmpty(MapDiff<K, T, H> diff) {
      */
     private static class JdkMapDiff<K, T> extends MapDiff<K, T, Map<K, T>> {
 
+        private JdkMapDiff(
+            KeySerializer<K> keySerializer,
+            ValueSerializer<K, T> valueSerializer,
+            List<K> deletes,
+            List<Map.Entry<K, Diff<T>>> diffs,
+            List<Map.Entry<K, T>> upserts
+        ) {
+            super(keySerializer, valueSerializer, deletes, diffs, upserts);
+        }
+
         private JdkMapDiff(StreamInput in, KeySerializer<K> keySerializer, ValueSerializer<K, T> valueSerializer) throws IOException {
             super(in, keySerializer, valueSerializer);
         }
 
-        private JdkMapDiff(Map<K, T> before, Map<K, T> after, KeySerializer<K> keySerializer, ValueSerializer<K, T> valueSerializer) {
-            super(keySerializer, valueSerializer);
+        private static <K, T> JdkMapDiff<K, T> create(
+            Map<K, T> before,
+            Map<K, T> after,
+            KeySerializer<K> keySerializer,
+            ValueSerializer<K, T> valueSerializer
+        ) {
             assert after != null && before != null;
 
             int inserts = 0;
-            for (Map.Entry<K, T> partIter : after.entrySet()) {
-                final K key = partIter.getKey();
-                T beforePart = before.get(key);
-                if (beforePart == null) {
-                    upserts.put(key, partIter.getValue());
+            var upserts = new ArrayList<Map.Entry<K, T>>();
```

Comment:
We could use `Tuple` instead of `Map.Entry` here couldn't we?

Reply:
Using `Map.Entry` allows to reuse entry returned by the `entrySet()` iterator while calculating the diff.
It is also less confusing to know where is the key and value.